File: internal/appwindow/panel_open_timing.go
package appwindow

import (
	"encoding/json"
	"os"
	"time"
)

type panelOpenTimingStage struct {
	Stage     string  `json:"stage"`
	ElapsedMS float64 `json:"elapsedMs"`
}

type panelOpenTiming struct {
	started                      time.Time
	now                          func() time.Time
	transferID, clusterID, phase string
	stages                       []panelOpenTimingStage
	log                          func(string, string)
}

// ConfigurePanelOpenTiming connects opt-in diagnostics to the existing app log.
// This is set by composition before windows are created, never during a transfer.
func (r *Registry) ConfigurePanelOpenTiming(log func(message, clusterID string)) {
	if os.Getenv("VITE_PANEL_OPEN_TIMING") == "1" {
		r.panelOpenTimingLog = log
	}
}

func newPanelOpenTiming(transferID, clusterID, phase string, log func(string, string)) *panelOpenTiming {
	if log == nil {
		return nil
	}
	return &panelOpenTiming{started: time.Now(), now: time.Now, transferID: transferID, clusterID: clusterID, phase: phase, log: log}
}

func (t *panelOpenTiming) mark(stage string) {
	if t != nil {
		t.stages = append(t.stages, panelOpenTimingStage{Stage: stage, ElapsedMS: float64(t.now().Sub(t.started)) / float64(time.Millisecond)})
	}
}

// Batch output after the operation returns so logging is outside transfer locks.
func (t *panelOpenTiming) report() {
	if t == nil {
		return
	}
	data, err := json.Marshal(struct {
		TransferID    string                 `json:"transferId"`
		Phase         string                 `json:"phase"`
		StartedUnixMS float64                `json:"startedUnixMs"`
		TotalMS       float64                `json:"totalMs"`
		Stages        []panelOpenTimingStage `json:"stages"`
	}{t.transferID, t.phase, float64(t.started.UnixMicro()) / 1000, float64(t.now().Sub(t.started)) / float64(time.Millisecond), t.stages})
	if err == nil {
		t.log("[DEBUG-panel-open] "+string(data), t.clusterID)
	}
}

File: frontend/src/core/panel-windows/panelOpenTiming.ts
// Opt-in diagnostics for a Windows development capture. Keep logging imports
// off the bootstrap path and emit batches after measuring, not at every mark.
type Stage = { stage: string; elapsedMs: number };
type Timing = {
  transferId: string;
  windowName: string;
  clusterId: string;
  started: number;
  stages: Stage[];
};

const enabled = import.meta.env.VITE_PANEL_OPEN_TIMING === '1';
const sources = new Map<string, Timing>();
const bootstrapStages: Stage[] = [];
let bootstrapReported = false;
let logger: Promise<typeof import('@/core/logging/appLogsClient')> | undefined;
const milliseconds = (value: number) => Math.round(value * 100) / 100;

function emit(timing: Timing, phase: 'source' | 'destination') {
  const message = `[DEBUG-panel-open] ${JSON.stringify({
    transferId: timing.transferId,
    windowName: timing.windowName,
    phase,
    startedUnixMs: performance.timeOrigin + timing.started,
    totalMs: milliseconds(performance.now() - timing.started),
    stages: timing.stages,
  })}`;
  logger ??= import('@/core/logging/appLogsClient');
  void logger
    .then(({ logAppLogsInfo }) =>
      logAppLogsInfo(message, 'PanelOpenTiming', { clusterId: timing.clusterId })
    )
    .catch(() => {
      // Diagnostic delivery must not affect a transfer or bootstrap failure.
    });
}

export function startPanelOpenTiming(transferId: string, clusterId: string, windowName: string) {
  if (enabled && !sources.has(transferId)) {
    sources.set(transferId, {
      transferId,
      clusterId,
      windowName,
      started: performance.now(),
      stages: [],
    });
  }
}

export function markPanelOpenTiming(transferId: string, stage: string) {
  const timing = sources.get(transferId);
  if (timing) {
    timing.stages.push({ stage, elapsedMs: milliseconds(performance.now() - timing.started) });
  }
}

export function finishPanelOpenTiming(transferId: string) {
  const timing = sources.get(transferId);
  if (timing) {
    markPanelOpenTiming(transferId, 'settled');
    sources.delete(transferId);
    emit(timing, 'source');
  }
}

export function markPanelBootstrapTiming(stage: string) {
  if (enabled && !bootstrapReported) {
    bootstrapStages.push({ stage, elapsedMs: milliseconds(performance.now()) });
  }
}

export function reportPanelBootstrapTiming(
  transferId: string,
  clusterId: string,
  windowName: string
) {
  if (enabled && !bootstrapReported) {
    bootstrapReported = true;
    emit({ transferId, clusterId, windowName, started: 0, stages: bootstrapStages }, 'destination');
  }
}
